nn_from_scratch/hardware/model_fc.c

```diff
@@ -0,0 +1,83 @@
+
+#include "model_fc.h"
+// std
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+// for print
+#include <stdio.h>
+
+#include "data/eqcheck_data.h"
+
+/*
+    if training flag is set and gradient is not NULL,
+              gradients will be updated for training
+
+*/
+
+float *fc_model(float *input_, int size, ModelPtr *model, int training_flag, float *predicted, Gradients *gradients)
+{
+    // forward propagate through each layer
+    float *input = malloc(size * sizeof(float));
+    memcpy(input, input_, size * sizeof(float)); // copy input, so the initial input does not get free'd
+    for (int i = 0; i < model->n_layers; i++)
+    {
+        float *output = fc_forward_prop(input, model->layers_weights[i], model->layers_biases[i],
+                                        size, model->layers_size[i], model->layers_activation[i], NULL);
+
+        if (!training_flag)
+        {
+            free(input);
+        }
+
+        input = output;
```

Comment:
Although not freeing up the output/activation, you are not tracking them (where is their address in memory to be used later on). Isn't that needed?